be/src/olap/rowset/segment_v2/segment_writer.cpp

```diff
@@ -0,0 +1,191 @@
+        RETURN_IF_ERROR(_index_builder->add_item(encoded_key));
+        _block_count++;
+    }
+    _row_count++;
+    return Status::OK();
+}
+
+template Status SegmentWriter::append_row(const RowCursor& row);
+
+uint64_t SegmentWriter::estimate_segment_size() {
+    return 0;
+}
+
+Status SegmentWriter::finalize(uint32_t* segment_file_size) {
+    for (auto column_writer : _column_writers) {
+        RETURN_IF_ERROR(column_writer->finish());
+    }
+    RETURN_IF_ERROR(_write_header());
+    RETURN_IF_ERROR(_write_data());
+    RETURN_IF_ERROR(_write_ordinal_index());
+    RETURN_IF_ERROR(_write_short_key_index());
+    RETURN_IF_ERROR(_write_footer());
+    return Status::OK();
+}
+
+// write header
+Status SegmentWriter::_write_header() {
+    std::string header_buf;
+    SegmentHeaderPB header;
+    if (!header.SerializeToString(&header_buf)) {
+        return Status::InternalError("failed to serialize segment header");
+    }
+
+    // header length
+    std::string magic_and_len_buf;
+    magic_and_len_buf.append(k_segment_magic);
+    put_fixed32_le(&magic_and_len_buf, header_buf.size());
+
+    // checksum
+    uint8_t checksum_buf[4];
+    // TODO(zc): add checksum
+
+    std::vector<Slice> slices = {magic_and_len_buf, header_buf, {checksum_buf, 4}};
+    RETURN_IF_ERROR(_write_raw_data(slices));
+    return Status::OK();
+}
+
+// write column data to file one by one
+Status SegmentWriter::_write_data() {
+    for (auto column_writer : _column_writers) {
+        RETURN_IF_ERROR(column_writer->write_data());
+    }
+    return Status::OK();
+}
+
+// write ordinal index after data has been written
+Status SegmentWriter::_write_ordinal_index() {
+    for (auto column_writer : _column_writers) {
+        RETURN_IF_ERROR(column_writer->write_ordinal_index());
+    }
+    return Status::OK();
+}
+
+Status SegmentWriter::_write_short_key_index() {
+    std::vector<Slice> slices;
+    // TODO(zc): we should get segment_size
+    RETURN_IF_ERROR(_index_builder->finalize(_row_count * 100, _row_count, &slices));
+
+    uint64_t offset = _output_file->size();
+    RETURN_IF_ERROR(_write_raw_data(slices));
+    uint32_t written_bytes = _output_file->size() - offset;
+
+    _footer.mutable_short_key_index_page()->set_offset(offset);
+    _footer.mutable_short_key_index_page()->set_size(written_bytes);
+    return Status::OK();
+}
+
+Status SegmentWriter::_write_footer() {
+    _footer.set_num_rows(_row_count);
+    // collect all 
+    for (int i = 0; i < _column_writers.size(); ++i) {
+        _column_writers[i]->write_meta(_footer.mutable_columns(i));
+    }
+
+    // write footer
+    std::string footer_buf;
+    if (!_footer.SerializeToString(&footer_buf)) {
+        return Status::InternalError("failed to serialize segment footer");
+    }
+    std::string magic_and_len_buf;
+    magic_and_len_buf.append(k_segment_magic, k_segment_magic_length);
+    put_fixed32_le(&magic_and_len_buf, footer_buf.size());
```

Comment:
we'd better follow the convention to write magic number after footer size

Reply:
OK